### ISSUE TITLE
YOLOv8 OBB - Out-of-bounds coordinates (<0 & >1)

### ISSUE THREAD

**Issue by YuriGagarine** [ultralytics/ultralytics#8747]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello everyone,

I decided to comment the assert lines which consider images with outside coordinate as corrupted images.
With my knowledge of YOLO, I do not understand such constraint. I think we can allow outside coordinate in a certain range for example from [0-eps; 1 + eps] with eps <1.

From this situation I have two questions :

- Did I just allowed the model to have an unexpected behaviour by commenting the lines ?
- If no, is it possible to, maybe, allow outside coordinate from a certain range ?

Thank you 

Hi everyone,

I recently commented out the assert lines in the code that flag images with out-of-bounds coordinates as corrupted. As someone new to YOLO, I'm not quite sure why this check is in place. It seems reasonable to allow coordinates to fall outside the image (expacially for OBB) boundaries by a small margin, say within the range of [0-ε, 1+ε], where ε is a small positive value less than 1.

Given this change, I have two questions:

-    Have I inadvertently introduced unexpected behavior by commenting out these lines?
-    If not, is it possible to modify the code to allow coordinates within a specific tolerance range around the image boundaries?

Thanks,

### Additional

The lines to comment to reproduce my experiment in ultralytics/data/utils.py in verify_image_label func:
```
assert points.max() <= 1, f"non-normalized or out of bounds coordinates {points[points > 1]}"
assert lb.min() >= 0, f"negative label values {lb[lb < 0]}"
```

**Comment by glenn-jocher** [comment]
@YuriGagarine hi there,

Great questions! By commenting out those assert lines, you're allowing coordinates outside the [0, 1] range, which typically represent normalized coordinates in the image space. Here's a quick rundown:

1. **Unexpected Behavior**: Allowing out-of-bounds coordinates might introduce unexpected behavior, especially in tasks like training or inference where the model expects coordinates to be within the image boundaries. This could potentially affect the model's ability to learn spatial relationships accurately.

2. **Modifying for Tolerance**: To allow a specific tolerance range, you could adjust the assert conditions to check for your ε tolerance. For example:

```python
epsilon = 0.05  # Define your tolerance
assert points.max() <= 1 + epsilon, f"coordinates exceed 1+epsilon {points[points > 1 + epsilon]}"
assert lb.min() >= -epsilon, f"coordinates below -epsilon {lb[lb < -epsilon]}"
```

This way, you're explicitly allowing a margin around the boundaries but still keeping a check to avoid extreme cases.

Hope this helps! 😊

**Comment by YuriGagarine** [comment]
Hi!

@glenn-jocher 
>     Allowing out-of-bounds coordinates might introduce unexpected behavior, especially in tasks like training or inference **where the model expects coordinates to be within the image boundaries**. This could potentially affect the model's ability to learn spatial relationships accurately.

I haven't found any specific code sections enforcing in-bound coordinates, except for the commented lines. I checked the Detect and OBB classes in head.py - they have a branch for detecting xywh format, but the output doesn't have any restrictions. Here's the relevant code snippet:
```
self.cv2 = nn.ModuleList(
    nn.Sequential(Conv(x, c2, 3), Conv(c2, c2, 3), nn.Conv2d(c2, 4 * self.reg_max, 1)) for x in ch
)
```

Similarly, the BBoxLoss and RotatedBboxLoss classes in loss.py don't seem to have any limitations on the coordinates.

If you have a chance, could you please clarify this further? I'd appreciate any insights you might have on this topic.

Thanks!

**Comment by glenn-jocher** [comment]
Hi @YuriGagarine,

You're right; the code snippets you've mentioned indeed don't explicitly enforce in-bound coordinates for detections. The model's architecture and loss functions are designed to work with the data as presented during training. The assertions you commented out serve as a data quality check rather than a model constraint. They help ensure that the training data fed into the model is normalized and within expected bounds, which is a common practice to maintain data integrity and model performance.

If your dataset inherently includes out-of-bounds coordinates that are meaningful for your specific use case, adjusting these checks, as you've done, is a reasonable approach. Just be mindful of how this might affect model training and inference, as the model learns from the data it's given. Keeping a close eye on model performance metrics during training and validation will be key to understanding the impact of this change. 🚀

Hope this clarifies things a bit!
